gno.land/pkg/sdk/vm/convert.go

```diff
@@ -207,3 +207,140 @@ func convertFloat(value string, precision int) float64 {
 
 	return f64
 }
+
+func JSONPrimitiveValues(m *gno.Machine, tvs []gno.TypedValue) string {
+	var str strings.Builder
+
+	str.WriteRune('[')
+	for i, tv := range tvs {
+		if i > 0 {
+			str.WriteRune(',')
+		}
+		str.WriteString(JSONPrimitiveValue(m, tv))
+	}
+	str.WriteRune(']')
+
+	return str.String()
+}
+
+func JSONPrimitiveValue(m *gno.Machine, tv gno.TypedValue) string {
+	if tv.T == nil {
+		return "null"
+	}
+
+	bt := gno.BaseOf(tv.T)
+	switch bt := bt.(type) {
+	case gno.PrimitiveType:
+		switch bt {
+		case gno.UntypedBoolType, gno.BoolType:
+			return fmt.Sprintf("%t", tv.GetBool())
+		case gno.UntypedStringType, gno.StringType:
+			return strconv.Quote(tv.GetString())
+		case gno.Float32Type:
+			f32 := math.Float32frombits(tv.GetFloat32())
+			return fmt.Sprintf("%f", f32)
+		case gno.Float64Type:
+			f64 := math.Float64frombits(tv.GetFloat64())
+			return fmt.Sprintf("%f", f64)
+		case gno.UntypedBigintType, gno.BigintType:
+			return tv.V.(gno.BigintValue).V.String()
+		case gno.UntypedBigdecType, gno.BigdecType:
+			return tv.V.(gno.BigdecValue).V.String()
+		case gno.IntType, gno.Int8Type, gno.Int16Type, gno.Int32Type, gno.UntypedRuneType, gno.Int64Type:
+			return fmt.Sprintf("%d", getSignedIntValue(bt, tv))
+		case gno.UintType, gno.Uint8Type, gno.Uint16Type, gno.Uint32Type, gno.Uint64Type, gno.DataByteType:
+			return fmt.Sprintf("%d", getUnsignedIntValue(bt, tv))
+		default:
+			panic("invalid primitive type - should not happen")
+		}
+	case *gno.ArrayType:
+		if bt.Elt == gno.Uint8Type {
+			arr := tv.V.(*gno.ArrayValue)
+			if data := arr.Data; data != nil {
+				i := arr.GetLength()
+				return `"` + base64.StdEncoding.EncodeToString(data[:i]) + `"`
+			}
+		}
+	case *gno.SliceType:
+		if bt.Elt == gno.Uint8Type {
+			slice := tv.V.(*gno.SliceValue)
+			if data := slice.GetBase(nil).Data; data != nil {
+				i := slice.GetLength()
+				return `"` + base64.StdEncoding.EncodeToString(data[:i]) + `"`
+			}
+		}
+	}
+
+	if tv.V == nil {
+		return "null"
+	}
+
+	if res, ok := tryGetError(m, tv); ok {
+		return res
+	}
+
+	// Fallback on generic object handler
+	if pv, ok := tv.V.(gno.PointerValue); ok {
```

Comment:
Could we not also consider the case where tv.V is an Object?

```
	switch v := tv.V.(type) {
	case *PointerValue:
	case Object:
	default:
	}
```

Reply:
tell me if it make sense this way: https://github.com/gnolang/gno/pull/2949/files#diff-11181752674b34519501331768f62e669f2babeb8f9b1afddda21e8465e3b874R282-R288